File: app/src/main/java/com/banglalink/toffee/model/ChannelInfo.java
package com.banglalink.toffee.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.banglalink.toffee.ui.player.HlsLinks;
import com.banglalink.toffee.util.Utils;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Created by MD.TAUFIQUR RAHMAN on 11/22/2016.
 */

public class ChannelInfo implements Parcelable {

    public String id;
    public String program_name;
    public String video_share_url;
    public String video_trailer_url;
    public String description;
    public String water_mark_url;
    public String type;
    public String view_count;
    public String formatted_view_count;
    public String lcn;
    public String individual_price;
    public String video_tags;
    public String duration;
    public String formattedDuration;
    public String age_restriction;
    public String service_operator_id;
    public String logo_mobile_url;
    public String poster_url_mobile;
    public boolean subscription;
    public boolean individual_purchase;
    public String expireTime;
    public List<HlsLinks> hlsLinks;
    public String channel_logo;
    public String category;
    public String subCategory;
    public int subCategoryId;
    public String favorite;
    public String potrait_ratio_800_1200;
    public String landscape_ratio_1280_720;
    public String feature_image;
    public String content_provider_name;
    public String content_provider_id;
    @SerializedName("url_type")
    public int urlType;


    public ChannelInfo() {

    }

    public boolean isLive(){
        return "LIVE".equalsIgnoreCase(type);
    }

    public boolean isVOD(){
        return "VOD".equalsIgnoreCase(type);
    }

    public boolean isCatchup(){
        return "CATCHUP".equalsIgnoreCase(type);
    }

    public String getHlsLink(){
        return hlsLinks.get(0).hls_url_mobile;
    }

//    protected ChannelInfo(ChannelInfo channelInfo){
//        this.id = channelInfo.id;
//        this.program_name = channelInfo.program_name;
//        this.video_share_url = channelInfo.video_share_url;
//        this.video_trailer_url = channelInfo.video_trailer_url;
//        this.description = channelInfo.description;
//        this.water_mark_url = channelInfo.water_mark_url;
//        this.type = channelInfo.type;
//        this.view_count = channelInfo.view_count;
//        this.lcn = channelInfo.lcn;
//        this.individual_price = channelInfo.individual_price;
//        this.video_tags = channelInfo.video_tags;
//        this.duration = channelInfo.duration;
//        this.age_restriction = channelInfo.age_restriction;
//        this.service_operator_id = channelInfo.service_operator_id;
//        this.logo_mobile_url = channelInfo.logo_mobile_url;
//        this.poster_url_mobile = channelInfo.poster_url_mobile;
//        this.subscription = channelInfo.subscription;
//        this.expireTime = channelInfo.expireTime;
//        this.hlsLinks = channelInfo.hlsLinks;
//        this.channel_logo = channelInfo.channel_logo;
//    }

    protected ChannelInfo(Parcel in) {
        id = in.readString();
        program_name = in.readString();
        video_share_url = in.readString();
        video_trailer_url = in.readString();
        description = in.readString();
        water_mark_url = in.readString();
        type = in.readString();
        view_count = in.readString();
        formatted_view_count = in.readString();
        lcn = in.readString();
        individual_price = in.readString();
        video_tags = in.readString();
        duration = in.readString();
        formattedDuration = in.readString();
        age_restriction = in.readString();
        service_operator_id = in.readString();
        logo_mobile_url = in.readString();
        poster_url_mobile = in.readString();
        subscription = in.readByte() != 0;
        individual_purchase = in.readByte() != 0;
        expireTime = in.readString();
        hlsLinks = in.createTypedArrayList(HlsLinks.CREATOR);
        channel_logo = in.readString();
        category = in.readString();
        subCategory = in.readString();
        subCategoryId = in.readInt();
        favorite = in.readString();
        potrait_ratio_800_1200 = in.readString();
        landscape_ratio_1280_720 = in.readString();
        feature_image = in.readString();
        content_provider_name = in.readString();
        content_provider_id = in.readString();
    }

    public static final Creator<ChannelInfo> CREATOR = new Creator<ChannelInfo>() {
        @Override
        public ChannelInfo createFromParcel(Parcel in) {
            return new ChannelInfo(in);
        }

        @Override
        public ChannelInfo[] newArray(int size) {
            return new ChannelInfo[size];
        }
    };

    public static List<HlsLinks> getHlsArrayList(List<String> links){
        List<HlsLinks> hlsLinks = new ArrayList<>();
        if(links != null){
            for(String link : links){
                HlsLinks hlsLink = new HlsLinks();
                hlsLink.hls_url_mobile = link;
                hlsLinks.add(hlsLink);
            }
        }
        return hlsLinks;
    }

    public static List<String> getStringArrayList(List<HlsLinks> hlsLinks){
        List<String> list = new ArrayList<>();
        if(hlsLinks != null){
            for(HlsLinks hls : hlsLinks){
                list.add(hls.hls_url_mobile);
            }
        }
        return list;
    }

    public String getCategory(String category, String subCategory){
        String itemCategory = "Channels>" + category;
        if (subCategory.length() > 0) {
            itemCategory += ">" + subCategory;
        }
        return itemCategory;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(program_name);
        dest.writeString(video_share_url);
        dest.writeString(video_trailer_url);
        dest.writeString(description);
        dest.writeString(water_mark_url);
        dest.writeString(type);
        dest.writeString(view_count);
        dest.writeString(formatted_view_count);
        dest.writeString(lcn);
        dest.writeString(individual_price);
        dest.writeString(video_tags);
        dest.writeString(duration);
        dest.writeString(formattedDuration);
        dest.writeString(age_restriction);
        dest.writeString(service_operator_id);
        dest.writeString(logo_mobile_url);
        dest.writeString(poster_url_mobile);
        dest.writeByte((byte) (subscription ? 1 : 0));
        dest.writeByte((byte) (individual_purchase ? 1 : 0));
        dest.writeString(expireTime);
        dest.writeTypedList(hlsLinks);
        dest.writeString(channel_logo);
        dest.writeString(category);
        dest.writeString(subCategory);
        dest.writeInt(subCategoryId);
        dest.writeString(favorite);
        dest.writeString(potrait_ratio_800_1200);
        dest.writeString(landscape_ratio_1280_720);
        dest.writeString(feature_image);
        dest.writeString(content_provider_name);
        dest.writeString(content_provider_id);
    }

    public boolean isPurchased() {
        return Integer.parseInt(individual_price) > 0 && individual_purchase;
    }

    public boolean isSubscribed(){
        return Integer.parseInt(individual_price
        ) == 0 && subscription;
    }

    public boolean isExpired(Date serverDate){
        try{
            return serverDate.after(Utils.getDate(expireTime));
        }catch (NullPointerException ne){
            return true;
        }
    }
}

File: app/src/main/java/com/banglalink/toffee/model/ChannelInfo.kt
package com.banglalink.toffee.model

import android.os.Parcelable
import android.text.Spanned
import android.util.Base64
import androidx.core.text.HtmlCompat
import com.banglalink.toffee.enums.Reaction
import com.banglalink.toffee.ui.player.HlsLinks
import com.banglalink.toffee.util.Utils
import com.banglalink.toffee.util.Utils.discardZeroFromDuration
import com.banglalink.toffee.util.UtilsKt
import com.banglalink.toffee.util.getFormattedViewsText
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.*

@Parcelize
data class ChannelInfo(
    var id: String,
    var program_name: String? = null,
    var video_share_url: String? = null,
    var video_trailer_url: String? = null,
    var description: String? = null,
    var water_mark_url: String? = null,
    var type: String? = null,
    var view_count: String? = null,
    var lcn: String? = null,
    var individual_price: String? = null,
    var video_tags: String? = null,
    var duration: String? = null,
    var age_restriction: String? = null,
    var service_operator_id: String? = null,
    var logo_mobile_url: String? = null,
    var poster_url_mobile: String? = null,
    var subscription:Boolean = false,
    var individual_purchase: Boolean = false,
    var expireTime: String? = null,
    var hlsLinks: List<HlsLinks>? = null,
    var channel_logo: String? = null,
    var category: String? = null,
    var subCategory: String? = null,
    var categoryId: Int = 0,
    var subCategoryId: Int = 0,
    var favorite: String? = null,
    var potrait_ratio_800_1200: String? = null,
    var landscape_ratio_1280_720: String? = null,
    var feature_image: String? = null,
    var content_provider_name: String? = null,
    var content_provider_id: String? = null,
    val channel_owner_id: Int = 0,
    var isSubscribed: Int = 0,
    var subscriberCount: Int = 0,

    @SerializedName("serial_name")
    val seriesName: String? = null,
    @SerializedName("total_season_no")
    val totalSeason: Int = 0,
    @SerializedName("season_no")
    val seasonNo: Int = 0,
    @SerializedName("serial_summary_id")
    val seriesSummaryId: Int = 0,
    @SerializedName("total_episode_no")
    val totalEpisode: Int = 0,
    @SerializedName("episode_no")
    val episodeNo: Int = 0,

    var is_available: Int = 0,
    var reaction: ReactionStatus? = null,   //individual reaction count from server
    var myReaction: Int = Reaction.None.value, //enum value (Reaction.Like.value) etc...
    var shareCount: Long = 0L,
    @SerializedName("playlist_content_id")
    val playlistContentId: Int = 0,
    @SerializedName("channel_profile_url")
    val channelProfileUrl: String? = null,
    @SerializedName("url_type")
    val urlType: Int = 0,
    @SerializedName("is_approved")
    val is_approved: Int? = null,
    val created_at: String? = null,
    @SerializedName("is_horizontal")
    val is_horizontal: Int? = null,
    @SerializedName("landscape_feature_1280_720")
    val ugcFeaturedImage: String? = null,
    @SerializedName("is_encoded")
    val isEncoded: Int? = null,
    @SerializedName("is_ugc")
    val is_ugc: Int = 0
) :Parcelable
{
    val isApproved: Int
        get() = if (is_approved == null || is_approved == 1) 1 else 0
    
    val isHorizontal: Int
        get() = if (is_horizontal == null || is_horizontal == 1) 1 else 0

    val isLive: Boolean
        get() = "LIVE".equals(type, ignoreCase = true)
    val isVOD: Boolean
        get() = "VOD".equals(type, ignoreCase = true)
    val isCatchup: Boolean
        get() = "CATCHUP".equals(type, ignoreCase = true)
    val isBucketUrl: Boolean
        get() = isEncoded == 0
    
    fun getCategory(category: String, subCategory: String): String {
        var itemCategory = "Channels>$category"
        if (subCategory.isNotEmpty()) {
            itemCategory += ">$subCategory"
        }
        return itemCategory
    }

    var viewProgress: Long = -1L
    fun viewProgressPercent(): Int {
        val durationInt = UtilsKt.getLongDuration(duration)
        if(viewProgress < 0L || durationInt <= 0L || isLive) return 0
        if(viewProgress > durationInt) return 1000
        return ((viewProgress.toDouble() / durationInt) * 1000L).toInt()
    }

    fun getDescriptionDecoded(): Spanned? {
        return try {
            HtmlCompat.fromHtml(String(Base64.decode(description, Base64.NO_WRAP)), HtmlCompat.FROM_HTML_MODE_LEGACY)
        } catch (ex: Exception) {
            null
        }
    }

    fun getHlsLink(): String? = hlsLinks?.get(0)?.hls_url_mobile

    fun formattedShareCount(): String = Utils.getFormattedViewsText(shareCount.toString())

    val isPurchased: Boolean
        get() = individual_price?.toInt() ?: 0 > 0 && individual_purchase
    val isPaidSubscribed: Boolean
        get() = individual_price?.toInt() == 0 && subscription

    fun isExpired(serverDate: Date): Boolean {
        return try {
            serverDate.after(Utils.getDate(expireTime))
        } catch (ne: NullPointerException) {
            true
        }
    }

    fun formattedViewCount(): String = getFormattedViewsText(view_count)
    fun formattedDuration(): String = discardZeroFromDuration(duration)
    fun formattedCreateTime(): String = if(!created_at.isNullOrBlank()) Utils.getDateDiffInDayOrHourOrMinute(Utils.getDate(created_at).time) else "0"
    fun formattedSubscriberCount(): String = getFormattedViewsText(subscriberCount.toString())
}